Resolve param indices using param values, not parameterset index

sadra-barikbin opened the pull request:
To compute param indices smarter when user passes multiple parameter sets with duplicate values to `parametrize` in order to have a better reordering experience. Currently we have :

```python
@pytest.fixture(scope='module')
def fixture1(request):
    pass

@pytest.fixture(scope='module')
def fixture2(request):
    pass

@pytest.mark.parametrize("fixture1, fixture2", [("a", 0), ("b", 1), ("a", 2)], indirect=True)
def test(fixture1, fixture2):
    pass
```
resulting in inefficient ordering i.e.
```
<Function test[a-0]>
<Function test[b-1]>
<Function test[a-2]>
```
It's a follow-up to #11220 .

sadra-barikbin committed:
Apply comments, rebase and a few improvements

sadra-barikbin committed:
Do the improvement

RonnyPfannschmidt commented:
@bluetech after some review of #519 and #2207 i believe it looks fair to consider the ordering more correct

i think we need a crosscheck of how we can observe interactions if people have multiple tests with different sets of parameters and their setup/teardown

bluetech commented:
@sadra-barikbin Can you please rebase this on latest main?

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Add a tiny test

sadra-barikbin committed:
Add a test

sadra-barikbin commented on src/_pytest/python.py:
```diff
@@ -1184,7 +1185,55 @@ def get_direct_param_fixture_func(request: FixtureRequest) -> Any:
+                argvalues_count[argname] += 1
+            except TypeError:  # `value` is not hashable
+                argname_indices.append(argvalues_count[argname])
+                argvalues_count[argname] += 1
```
I don't get why CodeCov considers this line not covered by tests.

RonnyPfannschmidt replied:
i believe there is a mistake in the try/except nesting - as we are already in the try for the key-error, so a type-error would fall out of it

this is a "how did this ever work" moment

RonnyPfannschmidt replied:
ok, now i understand the logic

ae5849f36281d7778f5f1aee89ece99586fa9c2d didn't add a testcase for that particular event, its indeed correctly handling as the errors originate in the first expression

sadra-barikbin committed:
Change a test

to use 'pytester.runpytest' instead of using testing/python/metafunc.py::TestMetafunc::Metafunc to see if coverage get fixed

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Remove a test

sadra-barikbin committed:
Revert a tiny change

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Remove a redundant test

sadra-barikbin committed:
Merge remote-tracking branch 'refs/remotes/upstream/Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

RonnyPfannschmidt commented:
Apologies but I'll be delayed in reviewing this, I'm still down with a cold

It would have been nice to land this in 8.x

sadra-barikbin commented:
> Apologies but I'll be delayed in reviewing this, I'm still down with a cold
> 
> It would have been nice to land this in 8.x

It's alright. I hope you recover soon.

RonnyPfannschmidt commented on src/_pytest/python.py:
```diff
@@ -1184,7 +1185,55 @@ def get_direct_param_fixture_func(request: FixtureRequest) -> Any:
+    :returns:
+        List of tuples of indices, each tuple for a parameter set.
+    """
+    indices = []
```
small nitpick, i believe `list[list[int]]` as type annotation will remove the need for cast

RonnyPfannschmidt reviewed (approved):
i think we have to document this better

from my pov we assign parameters that are used multiple times in a expanded/multiplied parameter set the index of the firs occurrence

this can be seen as approximation of a "multi dimensional" index from  filtered cross-product

sadra-barikbin committed:
Improve docstring and type annotation

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Fix a test

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin commented:
@bluetech , could you please review the PR too?

cyw233 commented:
Hi @sadra-barikbin, do we have any plan to merge this PR, please? cc. @RonnyPfannschmidt

obestwalter committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

obestwalter commented:
Just resolved a minor conflict. Let's see if we can get this reviewed and merged.

obestwalter commented:
Our shiny new bot also enforces changelog entries now, so this needs to be added before it can be merged - could you please add that @sadra-barikbin?

webknjaz commented:
@sadra-barikbin there's another merge conflict to be resolved in addition to the need of having a changelog fragment.

sadra-barikbin committed:
Resolve conflict; Add changelog

sadra-barikbin commented:
I resolved the conflict but this PR introduces is a breaking change with regard to how indices are assigned to the direct params of the test functions (hence one of the tests fail). Currently, when whole parametrizations of a function take place ,the indices of a direct param is assigned to its occurrences sequentially all at once:

https://github.com/pytest-dev/pytest/blob/f373974707f57a0b28d12563e4d03c7cd54c70d9/src/_pytest/python.py#L1452-L1456

To be able to merge this PR, I should bring the `resolve_values_indices_in_parametersets` logic in the method above which introduces change to the whole reordering.

Pierre-Sassoulas commented:
There seem to be a consensus that this should be merged but another conflict with the latest main appeared. Would you mind fixing it @sadra-barikbin ?

sadra-barikbin committed:
Merge branch 'main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Update the test

sadra-barikbin committed:
Merge remote-tracking branch 'origin/main' into Improvement-catch-duplicate-values-when-determining-param-indices-in-metafunc-parametrize

sadra-barikbin committed:
Add strict to zip

sadra-barikbin commented:
> There seem to be a consensus that this should be merged but another conflict with the latest main appeared. Would you mind fixing it @sadra-barikbin ?

This is breaking regarding the reordering behavior. Is it OK? It makes `test_high_scoped_parametrize_reordering` fail for example, **if I make it optimize the direct parametrizations as well.**

sadra-barikbin committed:
Fix the changelog

Zac-HD commented:
Closing this because the underlying issue has been fixed by https://github.com/pytest-dev/pytest/issues/14746; we group directly by values rather than using indices.
